### PR TITLE
Remove `dual-channel-replication` Feature Flag's Protection

### PR DESCRIPTION
**Pull request by naglera**
Currently, the `dual-channel-replication` feature flag is immutable if `enable-protected-configs` is enabled, which is the default behavior. This PR proposes to make the `dual-channel-replication` flag mutable, allowing it to be changed dynamically without restarting the cluster.

**Motivation:**
The ability to change the `dual-channel-replication` flag dynamically is essential for testing and validating the feature on real clusters running in production environments. By making the flag mutable, we can enable or disable the feature without disrupting the cluster's operations, facilitating easier testing and experimentation. Additionally, this change would provide more flexibility for users to enable or disable the feature based on their specific requirements or operational needs without requiring a cluster restart.

### CODE REVIEW

**Comment by madolson** [comment]
This title isn't right, it's removing that it's a protected config, which has wider implications that just making it a mutable config. Making it protected doesn't make sense, was there a reason @PingXie?

**Comment by PingXie** [comment]
> This title isn't right, it's removing that it's a protected config, which has wider implications that just making it a mutable config. Making it protected doesn't make sense, was there a reason @PingXie?

I don't remember the exact context now but I guess "protected" was introduced as a guard rail to prevent "fat-fingering" while keeping the config mutable. In a sense, this setting is kind of "semi-immutable". BTW, I have always been in favor of "immutable" but "mutable" was not a blocker to me either. 

I think two more things need to be done as part of this PR:

1. A description of the expected behavior. For example, what would happen if a user flips this config while a dual-channel full sync is in progress? 
3. a test suit that validates such a contract is upheld by the server.

**Comment by madolson** [comment]
> I don't remember the exact context now but I guess "protected" was introduced as a guard rail to prevent "fat-fingering" while keeping the config mutable. In a sense, this setting is kind of "semi-immutable". BTW, I have always been in favor of "immutable" but "mutable" was not a blocker to me either.

Ok, protected feels wrong to me, protected was intended for security configurations to make sure nobody can maliciously escalate their permissions. This feels like we aren't sure if something can be changed.

> A description of the expected behavior. For example, what would happen if a user flips this config while a dual-channel full sync is in progress?

I don't know what Amit had in mind, but it the sync should continue with full-sync. I don't think flipping these configs should abort the ongoing synchronization.

> a test suit that validates such a contract is upheld by the server.

What contract are you referring to?

File: tests/integration/dual-channel-replication.tcl [hunk 1]
```diff
@@ -251,6 +252,63 @@ start_server {tags {"dual-channel-replication external:skip"}} {
...
+                        }
+                    }
+                    # Toggle config
+                    set new_value "yes"
+                    if {$enable == "yes"} {
+                        set new_value "no"
+                    }
+                    set instance $primary
+                    if {$test_instance == "replica"} {
+                        set instnace $replica
```
Comment (naglera) [inline tests/integration/dual-channel-replication.tcl]:
```suggestion
                        set instance $replica
```

**Comment by madolson** [comment]
From core discussion, let's document that the sync finishes the replication with dual channel even when the flag is flipped part way through. (Ideally it should finish with the initial settings, but for now the scope is just dual channel).

**Comment by naglera** [comment]
Added tests to verify that sync continues normally in case of dual-channel config changed.

File: valkey.conf [hunk 2]
```diff
@@ -687,6 +687,12 @@ repl-diskless-load disabled
 # memory to accommodate the buffer during synchronization. However, this tradeoff is 
 # generally beneficial as it prevents potential performance degradation on the primary 
 # server, which is typically handling more critical operations.
+#
+# Additionally, it's important to note that when toggling this configuration on or off
```
Comment (zuiderkwast) [inline valkey.conf]: I wouldn't say "important to note". It's kind of expected behaviour and it's safe to toggle the config.

```suggestion
# When toggling this configuration on or off 
```

**Review by zuiderkwast — COMMENTED** [review]
Good. I didn't review the test.

File: tests/integration/dual-channel-replication.tcl [hunk 3]
```diff
@@ -251,6 +252,63 @@ start_server {tags {"dual-channel-replication external:skip"}} {
...
+
+                    $replica replicaof $primary_host $primary_port
+                    # wait for sync to start
+                    if {$enable == "yes"} {
+                        wait_for_condition 500 1000 {
+                            [string match "*slave*,state=wait_bgsave*,type=rdb-channel*" [$primary info replication]] &&
+                            [string match "*slave*,state=bg_transfer*,type=main-channel*" [$primary info replication]] &&
+                            [s 0 rdb_bgsave_in_progress] eq 1
+                        } else {
+                            fail "replica didn't start sync session in time"
```
Comment (PingXie) [inline tests/integration/dual-channel-replication.tcl]:
```suggestion
                            fail "replica didn't start a dual-channel sync session in time"
```

File: tests/integration/dual-channel-replication.tcl [hunk 4]
```diff
@@ -251,6 +252,63 @@ start_server {tags {"dual-channel-replication external:skip"}} {
...
+                        }
+                    } else {
+                        wait_for_condition 500 1000 {
+                            [string match "*slave*,state=wait_bgsave*,type=replica*" [$primary info replication]] &&
+                            [s 0 rdb_bgsave_in_progress] eq 1
+                        } else {
+                            fail "replica didn't start sync session in time"
```
Comment (PingXie) [inline tests/integration/dual-channel-replication.tcl]:
```suggestion
                            fail "replica didn't start a normal full sync session in time"
```

File: tests/integration/dual-channel-replication.tcl [hunk 5]
```diff
@@ -251,6 +252,63 @@ start_server {tags {"dual-channel-replication external:skip"}} {
...
+                    if {$enable == "yes"} {
+                        set new_value "no"
+                    }
+                    set instance $primary
+                    if {$test_instance == "replica"} {
+                        set instance $replica
+                    }
+                    $instance config set dual-channel-replication-enabled $new_value
+                    # Wait for at least one server cron
+                    after 1
```
Comment (PingXie) [inline tests/integration/dual-channel-replication.tcl]: maybe wait for a full second to make the case stronger?

```suggestion
                    after 1000
```

File: tests/integration/dual-channel-replication.tcl [hunk 6]
```diff
@@ -251,6 +252,63 @@ start_server {tags {"dual-channel-replication external:skip"}} {
                 }
                 $replica replicaof no one
             }
+
+            foreach test_instance {primary replica} {
+                $primary config set dual-channel-replication-enabled $enable
+                $replica config set dual-channel-replication-enabled $enable
+                test "Online toggle dual-channel-replication-enabled: $enable start, toggle on $test_instance" {
```
Comment (PingXie) [inline tests/integration/dual-channel-replication.tcl]: Will this string capture the test scenario?
```suggestion
                test "Online toggle dual-channel-replication-enabled on $test_instance, starting with '$enable'" {    
```

**Review by PingXie — APPROVED** [review]
LGTM. Nits only. Thanks @naglera

File: tests/integration/dual-channel-replication.tcl (hunk 5 above)
Comment (naglera) [reply tests/integration/dual-channel-replication.tcl]: Right, it was my intention

**Review by madolson — APPROVED** [review]

**Review by lilpoozie2005 — APPROVED** [review]

**Pull request merged**